crates/ethereum/engine/src/orchestrator.rs

```diff
@@ -0,0 +1,134 @@
+use reth_beacon_consensus::{BeaconEngineMessage, EthBeaconConsensus};
+use reth_chainspec::ChainSpec;
+use reth_db_api::database::Database;
+use reth_engine_tree::{
+    backfill::PipelineSync,
+    chain::ChainOrchestrator,
+    download::BasicBlockDownloader,
+    engine::{EngineApiEvent, EngineApiRequestHandler, EngineHandler, FromEngine},
+};
+use reth_ethereum_engine_primitives::EthEngineTypes;
+use reth_network_p2p::{bodies::client::BodiesClient, headers::client::HeadersClient};
+use reth_stages_api::Pipeline;
+use reth_tasks::TaskSpawner;
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tokio::sync::mpsc;
+use tokio_stream::wrappers::UnboundedReceiverStream;
+
+/// Alias for Ethereum chain orchestrator.
+type EthServiceType<DB, Client> = ChainOrchestrator<
+    EngineHandler<
+        EngineApiRequestHandler<EthEngineTypes>,
+        UnboundedReceiverStream<BeaconEngineMessage<EthEngineTypes>>,
+        BasicBlockDownloader<Client>,
+    >,
+    PipelineSync<DB>,
+>;
+
+/// The type that drives the Ethereum chain forward and communicates progress.
+#[pin_project]
+#[allow(missing_debug_implementations)]
+pub struct EthService<DB, Client>
+where
+    DB: Database + 'static,
+    Client: HeadersClient + BodiesClient + Clone + Unpin + 'static,
+{
+    orchestrator: EthServiceType<DB, Client>,
+}
+
+impl<DB, Client> EthService<DB, Client>
+where
+    DB: Database + 'static,
+    Client: HeadersClient + BodiesClient + Clone + Unpin + 'static,
+{
+    /// Constructor for `EthService`.
+    pub fn new(
+        chain_spec: Arc<ChainSpec>,
+        client: Client,
+        to_tree: mpsc::Sender<FromEngine<BeaconEngineMessage<EthEngineTypes>>>,
+        from_tree: mpsc::UnboundedReceiver<EngineApiEvent>,
+        incoming_requests: UnboundedReceiverStream<BeaconEngineMessage<EthEngineTypes>>,
+        pipeline: Pipeline<DB>,
+        pipeline_task_spawner: Box<dyn TaskSpawner>,
+    ) -> Self {
+        let consensus = Arc::new(EthBeaconConsensus::new(chain_spec));
+        let downloader = BasicBlockDownloader::new(client, consensus);
+
+        let engine_handler = EngineApiRequestHandler::new(to_tree, from_tree);
+        let handler = EngineHandler::new(engine_handler, downloader, incoming_requests);
+
+        let backfill_sync = PipelineSync::new(pipeline, pipeline_task_spawner);
+
+        Self { orchestrator: ChainOrchestrator::new(handler, backfill_sync) }
+    }
+}
+
+impl<DB, Client> Future for EthService<DB, Client>
+where
+    DB: Database + 'static,
+    Client: HeadersClient + BodiesClient + Clone + Unpin + 'static,
+{
+    type Output = Result<(), EthServiceError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        // Call poll on the inner orchestrator.
+        match ready!(self.project().orchestrator.poll_next_unpin(cx)) {
+            Some(_event) => Poll::Ready(Ok(())),
+            None => Poll::Ready(Ok(())),
+        }
```

Comment:
since this is intended as an endless future we should do a loop over the match and continue on Some(event) to not terminate the future

Reply:
sure, done